documentation/modules/oauth/proc-oauth-authorization-keycloak-example.adoc

```diff
@@ -0,0 +1,632 @@
+// Module included in the following module:
+//
+// assembly-oauth-authorization.adoc
+
+[id='proc-oauth-authorization-keycloak-example_{context}']
+= Trying Keycloak Authorization Services
+
+[role="_abstract"]
+This example explains how to use Keycloak Authorization Services with `keycloak` authorization.
+Use Keycloak Authorization Services to enforce access restrictions on Kafka clients.
+Keycloak Authorization Services use authorization scopes, policies and permissions to define and apply access control to resources.
+
+Keycloak Authorization Services REST endpoints provide a list of granted permissions on resources for authenticated users.
+The list of grants (permissions) is fetched from the Keycloak server as the first action after an authenticated session is established by the Kafka client
+The list is refreshed in the background so that changes to the grants are detected.
+Grants are cached and enforced locally on the Kafka broker for each user session to provide fast authorization decisions.
+
+Strimzi provides two example files with the deployment artifacts for setting up Keycloak:
+
+`kafka-ephemeral-oauth-single-keycloak-authz.yaml`:: An example `Kafka` custom resource configured for OAuth 2.0 token-based authorization using Keycloak.
+You can use the custom resource to deploy a Kafka cluster that uses `keycloak` authorization and token-based `oauth` authentication.
+
+`kafka-authz-realm.json`:: An example Keycloak realm configured with sample groups, users, roles and clients.
+You can import the realm into a Keycloak instance to set up fine-grained permissions to access Kafka.
+
+If you want to try Keycloak, use these files to perform the tasks  outlined in this section in the order shown.
+
+. xref:proc-oauth-authorization-keycloak-example-setup_{context}[Setting up the Keycloak Admin Console]
+. xref:proc-oauth-authorization-keycloak-example-deploy-kafka_{context}[Deploying a Kafka cluster with Keycloak authorization]
+. xref:proc-oauth-authorization-keycloak-example-authentication_{context}[Authenticating Keycloak access to Kafka brokers]
+. xref:proc-oauth-authorization-keycloak-example-check_{context}[Checking authorized access to Kafka using permissions]
+
+.Security considerations
+Keycloak authorization uses TLS connectivity where necessary. The server certificates are auto-generated. TLS termination for the Keycloak server is on a reverse proxy server using a Kubernetes ingress, not the Keycloak server pod. Take care when creating and reviewing any production configuration.
+Keycloak requires Kafka clients to use the same hostname and port as the Kafka brokers. This example assumes that the ingress hostname is visible to the pods inside the Kubernetes cluster. In your cluster, the ingress might be visible to external Kafka clients, but not to the pods. You can deploy the Keycloak server to a different Kubernetes cluster than the Cluster Operator.
```

Comment:
@mstruk I'm not sure I fully understand what we're trying to say with this section. Can we reduce it down and pull out the key point? For example, "In a production environment, you can deploy...to avoid..."

Reply:
Thanks for that, Marko. It made it a lot clearer. I've updated as follows:

**Security considerations**
Kafka and Kafka clients must connect to Keycloak for authorization to work. Connection to Keycloak is indirect. The connection routes through an ingress HTTP gateway using a hostname on port 443. Kafka clients use the same hostname and port as the Kafka brokers. The gateway is part of the Kubernetes infrastructure. This example assumes that Keycloak is deployed using the Keycloak Operator. The operator makes the gateway hostname visible to pods in the same Kubernetes cluster. The hostname must be visible for the example to work. If it's not visible, try deploying Keycloak to a separate Kubernetes cluster. The pods can then access Keycloak using the external hostname to make the connection.

Reply:
We could add a graphic if it helps

Reply:
I think it still needs a little bit of clarifying. This statement is incorrect:
'The operator makes the gateway hostname visible to pods in the same Kubernetes cluster.' 
Keycloak Operator can't make the gateway hostname visible to pods, it can only make use of the ingress. It can't ensure that the ingress hostname is visible to the pods - that part is a matter of Kubernetes administrator configuring the Kubernetes cluster to make the ingress hostname visible to the pods within that cluster.
'The operator exposes Keycloak over TLS using the ingress HTTP gateway. The ingress hostname must be visible for the example to work'. I don't think graphics is necessary, but if you think it is, sure.